### PR TITLE
Compute "executable_lines_data" metric

### LINKED ISSUE #53: Compute "executable_lines_data" metric
CoreMetrics.EXECUTABLE_LINES_DATA

Relates to [MMF-739](https://jira.sonarsource.com/browse/MMF-739)

### PR DESCRIPTION
**Pull request by alban-auzeill**
Fix #53

### CODE REVIEW

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/Metrics.java [hunk 1]
```diff
@@ -9,5 +9,6 @@
   public int numberOfStatements = 0;
 
   public Set<Integer> linesOfCode = new HashSet<>();
+  public Set<Integer> executableLines = new HashSet<>();
```
Comment (SonarTech) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/Metrics.java]: Vulnerability Vulnerability: Make executableLines a static final constant or non-public and provide accessors if needed. ([squid:ClassVariableVisibilityCheck](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=144&issues=AWJD5lYcxiwfozHKEGcR&open=AWJD5lYcxiwfozHKEGcR)

[ISSUE_KEY]: AWJD5lYcxiwfozHKEGcR

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/MetricsVisitor.java [hunk 2]
```diff
@@ -20,6 +21,7 @@ public void visitNode(UastNode node) {
     }
     if (node.kinds.contains(UastNode.Kind.STATEMENT)) {
       metrics.numberOfStatements++;
+      addLines(metrics.executableLines, node.firstToken());
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/MetricsVisitor.java]: It seems to me that it's not working for `switch` statement: Cases are not going to be highlighted.

The same problem will occur for other statements which could spread over multiple lines, and NOT having statement in their body. For instance, what about function invocations, if parameters are spread over multiple lines ?

```go
func foo(x,y,z int) {
  // ...
}

func bar() {
  foo(
    42 + 1 - 18,
    qix(42),
    myArray[42]
  )
}
```

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/MetricsVisitor.java [hunk 3]
```diff
@@ -32,8 +34,14 @@ public void visitToken(Set<UastNode.Kind> nodeKinds, UastNode.Token token) {
...
     }
     Set<Integer> lineNumbers = nodeKinds.contains(UastNode.Kind.COMMENT) ? metrics.commentLines : metrics.linesOfCode;
-    for (int line = token.line; line <= token.endLine; line++) {
-      lineNumbers.add(line);
+    addLines(lineNumbers, token);
+  }
+
+  private static void addLines(Set<Integer> lineNumbers, @Nullable UastNode.Token token) {
+    if (token != null) {
+      for (int line = token.line; line <= token.endLine; line++) {
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/MetricsVisitor.java]: What about using streams instead ?
```java
IntStream.rangeClosed(token.line, token.endLine).forEach(lineNumbers::add); 
```

File: sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoSensor.java [hunk 4]
```diff
@@ -93,10 +94,17 @@ private void saveMetrics(Metrics metrics, SensorContext context, InputFile input
...
     saveMetric(context, inputFile, CoreMetrics.STATEMENTS, metrics.numberOfStatements);
 
-    FileLinesContext fileLinesContext = fileLinesContextFactory.createFor(inputFile);
-    metrics.linesOfCode.forEach(line -> fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, 1));
-    metrics.commentLines.forEach(line -> fileLinesContext.setIntValue(CoreMetrics.COMMENT_LINES_DATA_KEY, line, 1));
-    fileLinesContext.save();
+    FileLinesContext linesContext = fileLinesContextFactory.createFor(inputFile);
+    saveLinesMetrics(linesContext, metrics.linesOfCode, CoreMetrics.NCLOC_DATA_KEY);
+    saveLinesMetrics(linesContext, metrics.commentLines, CoreMetrics.COMMENT_LINES_DATA_KEY);
+    if (inputFile.type() == InputFile.Type.MAIN) {
```
Comment (ghost) [inline sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoSensor.java]: I would maybe add a comment with a `FIXME` referencing  #140 here

**Review by ghost — CHANGES_REQUESTED** [review]
I let you have a look at my comments. My main concern is about missing executable lines when nested children of a given statement are executable, but are not identified as statement.

File: sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoSensor.java [hunk 5]
```diff
@@ -97,10 +98,18 @@ private void saveMetrics(Metrics metrics, SensorContext context, InputFile input
...
     saveMetric(context, inputFile, CoreMetrics.STATEMENTS, metrics.numberOfStatements);
 
-    FileLinesContext fileLinesContext = fileLinesContextFactory.createFor(inputFile);
-    metrics.linesOfCode.forEach(line -> fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, 1));
-    metrics.commentLines.forEach(line -> fileLinesContext.setIntValue(CoreMetrics.COMMENT_LINES_DATA_KEY, line, 1));
-    fileLinesContext.save();
+    FileLinesContext linesContext = fileLinesContextFactory.createFor(inputFile);
+    saveLinesMetrics(linesContext, metrics.linesOfCode, CoreMetrics.NCLOC_DATA_KEY);
+    saveLinesMetrics(linesContext, metrics.commentLines, CoreMetrics.COMMENT_LINES_DATA_KEY);
+    // FIXME currently *_test.go files are MAIN and not TEST, see issue #140
```
Comment (SonarTech) [inline sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoSensor.java]: Code Smell Code Smell: Take the required action to fix the issue indicated by this comment. ([squid:S1134](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=144&issues=AWJJA8_luc1UcxX5-Pdl&open=AWJJA8_luc1UcxX5-Pdl)

[ISSUE_KEY]: AWJJA8_luc1UcxX5-Pdl

Comment (SonarTech) [inline sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoSensor.java]: Code Smell Code Smell: Take the required action to fix the issue indicated by this comment. ([squid:S1134](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=144&issues=AWJJA8_luc1UcxX5-Pdl&open=AWJJA8_luc1UcxX5-Pdl)

[ISSUE_KEY]: AWJJA8_luc1UcxX5-Pdl

File: uast-generator-go/generate_source.go [hunk 6]
```diff
@@ -35,13 +35,14 @@ import (
...
 			NewTypeKind((*ast.AssignStmt)(nil), "t.computeAssignStmtKinds(astNode.Tok)..."),
 			NewTypeKind((*ast.BasicLit)(nil), "t.computeBasicLitKinds(astNode.Kind)..."),
-			NewTypeKind((*ast.ExprStmt)(nil), "EXPRESSION"),
 			NewTypeKind((*ast.BinaryExpr)(nil), "BINARY_EXPRESSION"),
 			NewTypeKind((*ast.CallExpr)(nil), "CALL"),
 			NewTypeKind((*ast.SelectorExpr)(nil), "SELECTOR_EXPR"),
 			NewTypeKind((*ast.SwitchStmt)(nil), "SWITCH"),
 			NewTypeKind((*ast.TypeSwitchStmt)(nil), "SWITCH"),
-			NewTypeKind((*ast.CaseClause)(nil), "CASE"),
+			NewTypeKind((*ast.CaseClause)(nil), "CASE, LABEL"),
```
Comment (ghost) [inline uast-generator-go/generate_source.go]: A `CASE` is not a `LABEL`. I would remove that change. Same for `CommClause`.

**Review by ghost — APPROVED** [review]
Except from LABEL which should be removed from kinds which are not labels... LGTM

**Review by SonarTech — COMMENTED** [review]
SonarQube analysis found issues:
Bug Bugs: 0
Vulnerability Vulnerabilities: 0
Code Smell Code Smells: 1

[See all issues in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=144&resolved=false)

File: sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoSensor.java [hunk 7]
```diff
@@ -98,10 +99,18 @@ private void saveMetrics(Metrics metrics, SensorContext context, InputFile input
...
```
Comment (SonarTech) [inline sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoSensor.java:105]: Code Smell Code Smell: Take the required action to fix the issue indicated by this comment. ([squid:S1134](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=144&issues=AWJJA8_luc1UcxX5-Pdl&open=AWJJA8_luc1UcxX5-Pdl)

[ISSUE_KEY]: AWJJA8_luc1UcxX5-Pdl

**Pull request merged**